doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,571 @@
+                (lambda r: 'fpga' in r.get('targets', []), FPGAPreprocessingPass()),
+            ]),
+            TaskletLanguageLoweringPass(),
+
+            # Phase 3: Code Generation
+            StateStructCreationPass(),
+            MemletLoweringPass(),
+            SDFGSplittingPass(),
+            FrameAndTargetCodeGenerationPass(),
+            HeaderGenerationPass(),
+        ])
+```
+
+### Target-Specific Sub-Pipelines
+
+```python
+class CUDACodegenPipeline(Pipeline):
+    """CUDA-specific code generation pipeline."""
+
+    def __init__(self):
+        super().__init__([
+            CUDATargetValidationPass(),
+            CUDAMemoryAnalysisPass(),
+            CUDAKernelExtractionPass(),
+            CUDAStreamAssignmentPass(),
+            CUDAKernelGenerationPass(),
+            CUDALaunchCodePass(),
+        ])
+
+class FPGACodegenPipeline(Pipeline):
+    """FPGA-specific code generation pipeline."""
+
+    def __init__(self):
+        super().__init__([
+            FPGAResourceAnalysisPass(),
+            FPGAStreamingPass(),
+            FPGAHLSGenerationPass(),
+            FPGAHostCodePass(),
+        ])
+```
+
+## Information Flow and Reuse
+
+### Pipeline Results Schema
+
+The `pipeline_results` dictionary will contain structured information to maximize reuse:
+
+```python
+pipeline_results = {
+    # From TypeInferencePass
+    "type_info": {
+        "connector_types": Dict[Node, Dict[str, dtypes.typeclass]],
+        "storage_types": Dict[str, dtypes.StorageType],
+        "schedule_types": Dict[Node, dtypes.ScheduleType],
+    },
+
+    # From MetadataCollectionPass
+    "metadata": {
+        "free_symbols": Dict[int, Set[str]],
+        "symbols_and_constants": Dict[int, Set[str]],
+        "arglist": Dict[str, data.Data],
+        "shared_transients": Dict[int, Set[str]],
+    },
+
+    # From AllocationAnalysisPass
+    "allocation_info": {
+        "to_allocate": DefaultDict[scope, List[allocation_tuple]],
+        "where_allocated": Dict[Tuple[SDFG, str], scope],
+        "allocation_scopes": Dict[str, scope],
+    },
+
+    # From TargetAnalysisPass
+    "targets": {
+        "required_targets": Set[TargetCodeGenerator],
+        "dispatch_routes": Dict[Node, TargetCodeGenerator],
+        "copy_dispatchers": Dict[edge, TargetCodeGenerator],
+    },
+
+    # From ControlFlowAnalysisPass
+    "control_flow": {
+        "control_flow_tree": Dict[SDFG, ControlFlow],
+        "structured_regions": Dict[SDFG, List[ControlFlowRegion]],
+    },
+
+    # From StreamAssignmentPass (GPU)
+    "stream_info": {
+        "stream_assignments": Dict[Node, int],
+        "stream_dependencies": nx.DiGraph,
+        "synchronization_points": List[Tuple[Node, Node]],
+    },
+
+    # From StateStructCreationPass
+    "state_struct": {
+        "struct_definition": str,
+        "struct_members": List[str],
+        "initialization_code": str,
+    },
+
+    # From AllocationCodePass
+    "allocation_code": {
```

Comment:
This needs to be clarified. There is no AllocationCodePass. There is an AllocationPass described above, which proposes tasklets instead of adding code to an `allocation_code` dictionary. Personally, I find the dictionary-based solution safer, at least as a first step in refactoring code generation.